### PR TITLE
Generate: fix end to end compilation

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

As @zucchini-nlp noticed, `RUN_SLOW=1 py.test tests/models/llama/test_modeling_llama.py::LlamaModelTest::test_generate_compile_fullgraph` was failing on `main`

This PR fixes it.

Ran locally:
- slow llama tests
- slow cache tests [NOTE: the torch.export test is failing, and was failing since it was introduced.]

### CODE REVIEW

File: src/transformers/cache_utils.py [hunk 1]
```diff
@@ -1024,19 +1024,22 @@ def __init__(self, config: PretrainedConfig, max_batch_size: int, max_cache_len:
...
+            #     breaks when updating the cache. It can't be used if the cache code is being compiled (but in that case
+            #     it is not needed anyway)
+            # 2. `torch.export()` requires mutations to be registered as buffers.
             if not is_torchdynamo_compiling():
-                torch._dynamo.mark_static_address(key_cache)
-                torch._dynamo.mark_static_address(value_cache)
-            self.key_cache.append(key_cache)
-            self.value_cache.append(value_cache)
+                self.register_buffer(f"key_cache_{idx}", torch.zeros(cache_shape, dtype=dtype, device=device))
+                self.register_buffer(f"value_cache_{idx}", torch.zeros(cache_shape, dtype=dtype, device=device))
```
Comment (gante) [inline src/transformers/cache_utils.py:1036]: `self.register_buffer` can't be compiled, but is needed for `torch.export`

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -1477,7 +1479,7 @@ def _get_cache(self, cache_implementation: str, max_batch_size: int, max_cache_l
                 "config": self.config,
                 "max_batch_size": max_batch_size,
                 "max_cache_len": max_cache_len,
-                "device": self.device,
+                "device": device,
```
Comment (gante) [inline src/transformers/generation/utils.py:1482]: I dropped this line in the end-to-end PR, in a rebase 😢 . `self.device` can't be called at compilation time

**Review by ArthurZucker — APPROVED** [review]
Let's ge tthis merged asap! I don't see tests updates so I suppose everything was caught by our CI but slow CI not push ci right?

**Comment by gante** [comment]
Correct, was caught by the slow ci :)

**Pull request merged**